shared-bindings/is31fl3741/FrameBuffer.c

```diff
@@ -0,0 +1,312 @@
+ * all copies or substantial portions of the Software.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+ * IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+ * FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+ * AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+ * LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+ * OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+ * THE SOFTWARE.
+ */
+
+#include "py/obj.h"
+#include "py/objproperty.h"
+#include "py/runtime.h"
+#include "py/objarray.h"
+
+#include "shared-bindings/is31fl3741/IS31FL3741.h"
+#include "shared-bindings/is31fl3741/FrameBuffer.h"
+#include "shared-bindings/util.h"
+#include "shared-module/displayio/__init__.h"
+#include "shared-module/framebufferio/__init__.h"
+#include "shared-module/framebufferio/FramebufferDisplay.h"
+#include "shared-bindings/busio/I2C.h"
+
+//| class IS31FL3741_FrameBuffer:
+//|     """Creates an in-memory framebuffer for a IS31FL3741 device."""
+//|
+//|     def __init__(self, is31: is31fl3741.IS31FL3741, width: int, height: int, mapping: Tuple[int, ...], *,
+//|         framebuffer: Optional[WriteableBuffer] = None, scale: bool = False, gamma: bool = False) -> None:
+//|         """Create a IS31FL3741_FrameBuffer object with the given attributes.
+//|
+//|         The framebuffer is in "RGB888" format using 4 bytes per pixel.
+//|         Bits 24-31 are ignored. The format is in RGB order.
+//|
+//|         If a framebuffer is not passed in, one is allocated and initialized
+//|         to all black.  In any case, the framebuffer can be retrieved
+//|         by passing the Is31fl3741 object to memoryview().
+//|
+//|         A Is31fl3741 is often used in conjunction with a
+//|         `framebufferio.FramebufferDisplay`.
+//|
+//|         :param is31fl3741.IS31FL3741 is31: base IS31FL3741 instance to drive the framebuffer
+//|         :param int width: width of the display
+//|         :param int height: height of the display
+//|         :param Tuple[int, ...] mapping: mapping of matrix locations to LEDs
+//|         :param Optional[WriteableBuffer] framebuffer: Optional buffer to hold the display
+//|         :param bool scale: if True display is scaled down by 3 when displayed
+//|         :param bool gamma: if True apply gamma correction to all LEDs"""
+//|         ...
+//|
+STATIC mp_obj_t is31fl3741_FrameBuffer_make_new(const mp_obj_type_t *type, size_t n_args, size_t n_kw, const mp_obj_t *all_args) {
+    enum { ARG_is31, ARG_width, ARG_height, ARG_framebuffer, ARG_mapping, ARG_scale, ARG_gamma };
+    static const mp_arg_t allowed_args[] = {
+        { MP_QSTR_is31, MP_ARG_OBJ | MP_ARG_REQUIRED | MP_ARG_KW_ONLY },
+        { MP_QSTR_width, MP_ARG_INT | MP_ARG_REQUIRED | MP_ARG_KW_ONLY },
+        { MP_QSTR_height, MP_ARG_INT | MP_ARG_REQUIRED | MP_ARG_KW_ONLY },
```

Comment:
Are these actually required to be keywords? They are documented like position arguments and used in the library as such. Probably best to just remove the `MP_ARG_KW_ONLY` here.

Reply:
I removed it from the mapping as well (also required) and reorged everything so it was in the right order.